### PR TITLE
feat(sql): add cumulative sum and first value window funtions

### LINKED ISSUE #3936: Incorrect result with window function
### Describe the bug

At some cases, the results are not expected.

`SELECT avg(T1.c_1) OVER(PARTITION BY 1=1 ORDER BY T1.c_0) FROM aaa AS T1 CROSS JOIN aaa AS T2;`

### To reproduce

```
create table aaa (c_0 timestamp, c_1 int);
insert into aaa values (NOW(), 1);
insert into aaa values (NOW(), 2);
```
```
SELECT avg(T1.c_1) OVER(PARTITION BY 1=1 ORDER BY T1.c_0) FROM aaa AS T1 CROSS JOIN aaa AS T2;
```
Result:
```
(1.0,)
(1.0,)
(1.5,)
(1.3333333333333333,)
```

### Expected Behavior

```
(1.0,)
(1.0,)
(1.5,)
(1.5,)
```

### Environment

```markdown
- **QuestDB version**: 7.3.4
- **OS**:
- **Browser**:
```

### Additional context

_No response_

### LINKED ISSUE #4010: Incorrect results with nested joins
### Describe the bug

Incorrect result. It does not show in previous version.

### To reproduce

```
create table t(c0 timestamp, c1 int, c2 int);
insert into t values('2023-09-21T10:00:00.000000Z',1,1);
insert into t values('2023-09-21T10:00:00.000000Z',1,1);
```
```
select count(*) from t as t1 join t as t2 on t1.c0<t2.c0 cross join t as t3;
```
result: 2

### Expected Behavior

expected result: 0

### Environment

```markdown
- **QuestDB version**: v7.3.5
- **OS**:
- **Browser**:
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by bziobrowski**
PR adds implementation of `sum()` and `first_value()` window  functions.

Example:
```sql
create table wintest
(
  ts timestamp,
  key symbol, 
  value long
) timestamp(ts) partition by month;

insert into wintest
select cast(x as timestamp), rnd_symbol('a', 'b', 'c'), x
from long_sequence(10);

 select ts, key, value, 
       sum(value) over (partition by key order by ts), 
       sum(value) over (partition by key rows between 1 preceding and current row ),
       sum(value) over (partition by key order by ts range between 1 microsecond preceding and current row ),
       first_value(value) over (partition by key order by ts), 
       first_value(value) over (partition by key rows between 1 preceding and current row ),
       first_value(value) over (partition by key order by ts range between 1 microsecond preceding and current row )
from wintest;

ts	key	value	sum	sum1	sum2	first_value	first_value1	first_value2
1970-01-01T00:00:00.000001Z	a	1	1	1	1	1	1	1
1970-01-01T00:00:00.000002Z	c	2	2	2	2	2	2	2
1970-01-01T00:00:00.000003Z	b	3	3	3	3	3	3	3
1970-01-01T00:00:00.000004Z	c	4	6	6	4	2	2	4
1970-01-01T00:00:00.000005Z	a	5	6	6	5	1	1	5
1970-01-01T00:00:00.000006Z	b	6	9	9	6	3	3	6
1970-01-01T00:00:00.000007Z	b	7	16	13	13	3	6	6
1970-01-01T00:00:00.000008Z	a	8	14	13	8	1	5	8
1970-01-01T00:00:00.000009Z	c	9	15	13	9	2	4	9
1970-01-01T00:00:00.000010Z	a	10	24	18	10	1	8	10

```

Fixes #4010 
Fixes #3936

### CODE REVIEW

**Comment by questdb-butler** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Please update the PR title to match this format:
`type(subType): description`

Where `type` is one of:
`feat`, `fix`, `chore`, `docs`, `style`, `refactor`, `perf`, `test`, `ci`, `revert`

And: `subType` is one of:
`build`, `sql`, `log`, `mig`, `core`, `ilp`, `pgwire`, `http`, `conf`, `ui`, `wal`

For Example:

```
perf(sql): improve pattern matching performance for SELECT sub-queries
```
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 5166bce88b5e4b628940aef70776d2ffe43f21a4
</p>

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 273 / 309 (88.35%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java|237|270|87.78%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/AvgDoubleWindowFunctionFactory.java|35|38|92.11%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|

File: core/src/main/java/io/questdb/griffin/FunctionParser.java [hunk 1]
```diff
@@ -713,7 +713,7 @@ private Function createFunction(
 
                 if (factory.isWindow()) {
                     // prefer window functions in window context, otherwise non-window functions
-                    sigArgTypeScore += isWindowContext ? -1 : 1;
```
Comment (bziobrowski) [inline core/src/main/java/io/questdb/griffin/FunctionParser.java:716]: I increased the value because function parser kept choosing regular SumLong..GroupByFunction over SumDoubleWindow for expressions like  `sum(1)`.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 287 / 311 (92.28%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/AvgDoubleWindowFunctionFactory.java|36|39|92.31%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java|250|271|92.25%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java [hunk 2]
```diff
@@ -0,0 +1,757 @@
...
+    public Function newInstance(
+            int position,
+            ObjList<Function> args,
+            IntList argPositions,
+            CairoConfiguration configuration,
+            SqlExecutionContext sqlExecutionContext
+    ) throws SqlException {
+        final WindowContext windowContext = sqlExecutionContext.getWindowContext();
+        if (windowContext.isEmpty()) {
+            throw SqlException.emptyWindowContext(position);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java:70]: how realistic we get this condition? This is untested

File: core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java [hunk 3]
```diff
@@ -0,0 +1,757 @@
...
+                    throw SqlException.$(windowContext.getRowsHiKindPos(), "frame end supports UNBOUNDED FOLLOWING only when frame start is UNBOUNDED PRECEDING");
+                }
+            }
+        }
+
+        int exclusionKind = windowContext.getExclusionKind();
+        int exclusionKindPos = windowContext.getExclusionKindPos();
+        if (exclusionKind != WindowColumn.EXCLUDE_NO_OTHERS
+                && exclusionKind != WindowColumn.EXCLUDE_CURRENT_ROW) {
+            throw SqlException.$(exclusionKindPos, "only EXCLUDE NO OTHERS and EXCLUDE CURRENT ROW exclusion modes are supported");
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java:93]: untested

File: core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java [hunk 4]
```diff
@@ -0,0 +1,757 @@
...
+        }
+
+        if (exclusionKind == WindowColumn.EXCLUDE_CURRENT_ROW) {
+            // assumes frame doesn't use 'following'
+            if (rowsHi == Long.MAX_VALUE) {
+                throw SqlException.$(exclusionKindPos, "EXCLUDE CURRENT ROW not supported with UNBOUNDED FOLLOWING frame boundary");
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java:99]: untested

Comment (bluestreak01) [reply core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java:99]: please review test coverage around edge conditions

File: core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java [hunk 5]
```diff
@@ -0,0 +1,757 @@
...
+        }
+
+        @Override
+        public void reset() {
+            map.close();
+        }
+
+        @Override
+        public void toPlan(PlanSink sink) {
+            sink.val(NAME);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java:345]: untested

File: core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java [hunk 6]
```diff
@@ -0,0 +1,757 @@
...
+            externalSum = Double.NaN;
+            count = 0;
+            sum = 0.0;
+        }
+
+        @Override
+        public void toTop() {
+            super.toTop();
+
+            externalSum = Double.NaN;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java:746]: untested

File: core/src/main/java/io/questdb/griffin/engine/functions/window/AvgDoubleWindowFunctionFactory.java [hunk 7]
```diff
@@ -611,7 +615,14 @@ public void computeNext(Record record) {
                     firstIdx = newFirstIdx;
                 }
 
-                avg = (frameSize != 0 ? sum / frameSize : Double.NaN);
+                if (frameSize != 0) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/functions/window/AvgDoubleWindowFunctionFactory.java:618]: because avg() is updated in this PR, i consider bugs in `avg()` relevant to this PR too :)

`close()` method is a bit of spaghetti, it is never called and instead memory is closed in `reset()`. Could you review this flow and make sure close is indeed called

File: core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java (hunk 6 above)
Comment (bluestreak01) [reply core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java:746]: what i mean is this:
image

File: core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java (hunk 2 above)
Comment (bziobrowski) [reply core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java:70]: Other functions can be tested for this condition with sql missing over clause but `avg()` and `sum()` get interpreted as usual group by functions. I reckon we can leave this just in case something goes wrong in function parser.

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 288 / 311 (92.60%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java|250|271|92.25%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/AvgDoubleWindowFunctionFactory.java|37|39|94.87%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/functions/window/AvgDoubleWindowFunctionFactory.java (hunk 7 above)
Comment (bziobrowski) [reply core/src/main/java/io/questdb/griffin/engine/functions/window/AvgDoubleWindowFunctionFactory.java:618]: Sure, thank you !

I inherited `reset()` from previous CachedWindowRecordCursorFactory and do agree it's confusing.
`reset()`
At the moment `close()` is only called for window functions when optimized path fails in SqlCodeGenerator . 
Should I replace `reset()` with `close()` ?

File: core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java (hunk 2 above)
Comment (bziobrowski) [reply core/src/main/java/io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java:70]: I kept the code and added artificial test to WindowFunctionTest .

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 794 / 869 (91.37%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/AvgDoubleWindowFunctionFactory.java|53|61|86.89%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/FirstValueDoubleWindowFunctionFactory.java|477|531|89.83%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/BaseDoubleWindowFunction.java|15|16|93.75%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java|220|232|94.83%|
|:large_blue_circle:|io/questdb/griffin/engine/join/CrossJoinRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/BasePartitionedDoubleWindowFunction.java|26|26|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/functions/window/AvgDoubleWindowFunctionFactory.java (hunk 7 above)
Comment (bluestreak01) [reply core/src/main/java/io/questdb/griffin/engine/functions/window/AvgDoubleWindowFunctionFactory.java:626]: I would let you decide, they should be one place and if you pick `reset` it would be good to have comments as to why it is enough to release memory there :)

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 830 / 905 (91.71%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/AvgDoubleWindowFunctionFactory.java|53|61|86.89%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/FirstValueDoubleWindowFunctionFactory.java|477|531|89.83%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/BaseDoubleWindowFunction.java|15|16|93.75%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java|220|232|94.83%|
|:large_blue_circle:|io/questdb/griffin/engine/AbstractRedBlackTree.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/CrossJoinRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTreeChain.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/BasePartitionedDoubleWindowFunction.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/window/CachedWindowRecordCursorFactory.java|1|1|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 827 / 901 (91.79%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/AvgDoubleWindowFunctionFactory.java|43|50|86.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/FirstValueDoubleWindowFunctionFactory.java|477|531|89.83%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/BaseDoubleWindowFunction.java|15|16|93.75%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java|220|232|94.83%|
|:large_blue_circle:|io/questdb/griffin/engine/AbstractRedBlackTree.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/CrossJoinRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTreeChain.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/BasePartitionedDoubleWindowFunction.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/window/CachedWindowRecordCursorFactory.java|1|1|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/functions/window/AvgDoubleWindowFunctionFactory.java (hunk 7 above)
Comment (bziobrowski) [reply core/src/main/java/io/questdb/griffin/engine/functions/window/AvgDoubleWindowFunctionFactory.java:622]: Turns out `reset()` and `close()` have slightly different purpose. 
`reset()` is used to release the native memory but keep function usable, so it keeps e.g. argument and partition keys. 
Once `close()` is called (on factory close), function can't be used anymore .

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 859 / 924 (92.97%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/AvgDoubleWindowFunctionFactory.java|59|67|88.06%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/FirstValueDoubleWindowFunctionFactory.java|486|531|91.53%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/BaseDoubleWindowFunction.java|18|19|94.74%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/SumDoubleWindowFunctionFactory.java|221|232|95.26%|
|:large_blue_circle:|io/questdb/griffin/engine/AbstractRedBlackTree.java|11|11|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/join/CrossJoinRecordCursorFactory.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/griffin/FunctionParser.java|10|10|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/window/WindowRecordCursorFactory.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTreeChain.java|24|24|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/functions/window/BasePartitionedDoubleWindowFunction.java|26|26|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/window/CachedWindowRecordCursorFactory.java|1|1|100.00%|

**Review by bluestreak01 — APPROVED** [review]
